### ISSUE TITLE
Expressify context

### ISSUE THREAD

**Issue by asbuch** [Nemocas/AbstractAlgebra.jl#1239]
This is a question, not an issue.
Is there a way to set custom properties of the context object received by `expressify()`?
I tried `Base.active_repl.options.iocontext[:mysymbol] = "myvalue"`, but I think `expressify()` uses a different context.
Thanks a lot!
Anders

**Comment by thofma** [comment]
Hm, in principle, this should work. Since the printing is quite convoluted, do you mind sharing what the end goal is?

**Comment by asbuch** [comment]
Many combinatorial objects have several different representations that are useful in different situations. I would like to be able to set my preferred representation globally. The printing context seems like a natural place, since then it can be overridden by `println(io, my_comb_obj)`. It would be natural if the context object received by `expressify()` would inherit any symbols set in the `iocontext` of `active_repl`, or in an `io` object passed to `println()`, or at least make the `io` object available.

**Comment by tthsqe12** [comment]
I am not sure how useful expressify could be in printing combinatorial objects. It is used currently for changing `x + -1*y` into `x-y` and for minimizing the parentheses in precedence correct printing.

**Comment by thofma** [comment]
What I do not understand yet is the relation to expressify. Since expressify for types defined in AbstractAlgebra cannot handle arbitrary properties that are stored in the IOContext, or that are passed on via the context, I gather this is about printing your own custom type? Then you define `Base.show(io::IO, ::MyType)` and can do with the IO context whatever you want.

**Comment by thofma** [comment]
Hm, the printing works as follows. For AbstractAlgebra types, where printing is done via expressify, it is roughly as follows:

```
show(io, x) ->
  show_via_expressify(io, x) ->
    println(io, obj_to_string(canonicalize(expressify(x, context))))
```

So there is no hook from `expressify` back into `show`.

So maybe the following could work:

```julia
function expressify(M::MyObject; context = context)
  build the expression
  opt = get(context, :youroption, false)
  if opt
    arg = [2, 1]
  else
    arg = [2, 3, 4, 5]
  end
  return Expr(:call, :*, 3, Expr(:call, :SchuberClass, arg))
end

function show(io::IOContext, M::MyObject)
  show_via_expressify(io::IO, MIME("text/plain"), M, context = io)
end
```

@tthsqe12 Can you verify that this would work in principle?

**Comment by asbuch** [comment]
This is what I was hoping to do! But my `expressify()` function usually receives the default value `nothing` in the `context` argument, so `get(context, :youroption, false)` will not work. Thanks for the source reference, will study.

**Comment by thofma** [comment]
Here is a working example. The `dict` should hold the options.
```
julia> struct MyType end

julia> function AbstractAlgebra.expressify(M::MyType; context = context)
         @show typeof(context.dict)
         opt = get(context, :youroption, false)
         if opt
           arg = [2, 1]
         else
           arg = [2, 3, 4, 5]
         end
         return Expr(:call, :*, 3, Expr(:call, :SchuberClass, AbstractAlgebra.expressify(arg)))
       end

julia> function Base.show(io::IOContext, M::MyType)
         AbstractAlgebra.show_via_expressify(io::IO, MIME("text/plain"), M, context = io)
       end

julia> MyType()
typeof(context.dict) = Base.ImmutableDict{Symbol, Any}
3*SchuberClass([2, 3, 4, 5])
```

**Comment by asbuch** [comment]
Thanks so much! This demonstrates that everything works like I hoped, but I must have made incorrect conclusions from my experiments. Issue solved!

**Comment by asbuch** [comment]
Actually, my `CohomElem` class is a subclass of `RingElement`, so it might be used as elements in a matrix or coefficients in a polynomial, in which case I don't think the trick will work. But it is still a lot better than nothing.

**Comment by thofma** [comment]
Yes, this will be a tad more tricky. Is there a specific reason why you want to manually adjust the REPL IO context and use that to control the printing? Alternatively (and equivalently) one could just have a global variable flying around. The advantage is, that this would be accessible in `expressify(::CohomElem)`.

**Comment by asbuch** [comment]
Using the print context would be more elegant, and would allow overriding the settings from a print command inside a function with something like:
`print(IOContext(stdout, :myopt => true), schub_class)`

But I will probably have to mess with global variables anyway, since `print(schub_class)` does not appear to pass options set in `Base.active_repl.options.iocontext` onto the `show()` function.

**Comment by tthsqe12** [comment]
> @tthsqe12 Can you verify that this would work in principle?
I would say this is good because use passed `AbstractAlgebra.expressify(arg)`. "Usual" julia objects don't work in the printer:
```
julia> AbstractAlgebra.show_obj(IOContext(stdout, :size_limit => 10), MIME("text/latex"), Expr(:call, :+, Expr(:call, :f, :x), collect(1:20)))
f\left(x\right) + [??? unknown object ???]
```
By passing `collect(1:20)` through `expressify`, it get the default treatment and is converted to a string.
```
julia> AbstractAlgebra.show_obj(IOContext(stdout, :size_limit => 10), MIME("text/latex"), Expr(:call, :+, Expr(:call, :f, :x), AbstractAlgebra.expressify(collect(1:20))))
f\left(x\right) + [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]
```
If you do the proper thing and expressify the array using `Expr(:vect, ...)`, then things like `size_limit` actually have something to work with.
```
julia> AbstractAlgebra.show_obj(IOContext(stdout, :size_limit => 10), MIME("text/latex"), Expr(:call, :+, Expr(:call, :f, :x), Expr(:vect, 1:20...)))
f\left(x\right) + \left[1, 2, 3, 4, 5, {\ldots}, 17, 18, 19, 20\right]
```
